File: tree/binary.js
function Node(value) {
  this.value = value
  this.left = null
  this.right = null
}

Node.prototype.visit = function () {
  if (this.left != null) {
   this.left.visit();
  }
  console.log(this.value);
  if (this.right != null) {
    this.right.visit();
  }
  return null;
}
Node.prototype.search = function (value) {
  if (this.value == value) {
    return this;
  } else if (value < this.value && this.left != null) {
    return this.left.search(value);
  } else if (value > this.value && this.right != null) {
    return this.right.search(value);
  }
}
Node.prototype.addNode = function (n) {

  if (n.value < this.value) {

    if (this.left == null) {
      this.left = n;
    } else {
      this.left.addNode(n)
    }
  } else {
    if (this.right == null) {
      this.right = n;
    } else {
      this.right.addNode(n)
    }
  }
}

function Tree() {
  this.root = null
}

Tree.prototype.addValue = function (value) {
  if (this.root == null) {
    this.root = new Node(value);
  } else {
    this.root.addNode(new Node(value));
  }
}
Tree.prototype.traverse = function (value) {
  this.root.visit();
}

Tree.prototype.search = function (value) {
  let found =   this.root.search(value);
  if (found != null){
    console.log(found)
  }
  return found
}
var tree;
tree = new Tree()
tree.addValue(123)
tree.addValue(12)
tree.addValue(22)
tree.addValue(23)
tree.traverse()

console.log(tree.search(23))
//console.log(tree.search(2))

File: tree/binary.py
class Node:
  def __init__(self, value):
    self.value = value
    self.left = None
    self.right = None

  def add_node(self, n):
    if n.value < self.value:
      if self.left == None:
        self.left = n
      else:
        self.left.add_node(n)

    else:
      if self.right == None:
        self.right = n
      else:
        self.right.add_node(n)

  def visit(self):
    if self.left:
      self.left.visit()
    print(self.value)
    if self.right:
      self.right.visit()

  def search(self, value):
    if self.value == value:
      return self

    elif value < self.value and self.left:
      return self.left.search(value)

    elif value > self.value and self.right:
      return self.right.search(value)

  def __repr__(self):
    return f"value={self.value}, left={self.left.value if self.left else None}, right={self.right.value if self.right else None}"


class Tree:
  def __init__(self):
    self.root = None

  def add_value(self, value):
    if not (self.root):
      self.root = Node(value)
    else:
      self.root.add_node(Node(value))

  def traverse(self):
    self.root.visit()

  def search(self, value):
    found = self.root.search(value)

    return found


tree = Tree()
tree.add_value(123)
tree.add_value(12)
tree.add_value(22)
tree.add_value(23)

tree.traverse()
print(tree.search(22))
